### PR TITLE
command: terraform providers mirror

### PR DESCRIPTION
**Pull request by apparentlymart**
This new command is intended to make it easy to create or update a mirror directory containing suitable providers for the current configuration, producing a layout that is appropriate both for a filesystem mirror or, if copied into the document root of an HTTP server, a network mirror. (Network mirrors are not yet supported by Terraform for installation, so _that_ part is more aspirational but the directory structures are compatible in any case.)

This initial version is not customizable aside from being able to select multiple platforms to install packages for.

Future iterations of this could include commands to turn the JSON index generation on and off, or to instruct it to produce the unpacked directory layout instead of the packed directory layout as it currently does. Both of those options would make the generated directory unsuitable to be a network mirror, but it would still work as a filesystem mirror.

In the long run this will hopefully form part of a replacement workflow to `terraform-bundle` as a way to put copies of providers somewhere so we don't need to re-download them every time, but some other changes will be needed outside of just this command before that'd be true, such as adding support for network and/or filesystem mirrors in Terraform Enterprise.

(I originally wrote this as a way to make it easier for me to test filesystem and network mirrors during development, so I had originally intended it to land as part of adding support for network mirrors. However, in retrospect it seems useful in already as a helper for building _filesystem_ mirrors, so I'm submitting it now to avoid it getting lost and forgotten in a random branch.)

### CODE REVIEW

File: command/e2etest/providers_mirror_test.go [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+	"testing"
+
+	"github.com/google/go-cmp/cmp"
+	"github.com/hashicorp/terraform/e2e"
+)
+
+// The tests in this file are for the "terraform providers schema" command,
+// which is tested in an e2etest mode rather than a unit test mode because it
+// interacts directly with Terraform Registry and the full details of that are
+// tricky to mock. Such a mock is _possible_, but we're using e2etest as a
```
Comment (mildwonkey) [inline command/e2etest/providers_mirror_test.go:17]: I am not at all opposed to this e2e test, but just so you know we recently added a `testRegistrySource` helper function in the command package:

https://github.com/hashicorp/terraform/blob/020084f6d044702d3a27dbc16800782e2bedb50e/command/command_test.go#L919

Comment (apparentlymart) [reply command/e2etest/providers_mirror_test.go:17]: This particular one is expecting to be talking always to upstream registries (so you don't have to unconfigure your mirrors just to populate them), so it wasn't very amenable to that sort of mocking. I did have an earlier version that created a mock registry on a fake domain and forced use of it via service discovery, but that required the fake providers to be properly signed and that was a little too complex to set up in a reasonable amount of time to get this done right now. Maybe in a later commit!

File: website/docs/commands/providers/mirror.html.md [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+[explicit installation method configuration](../cli-config.html#explicit-installation-method-configuration)
+allows you to configure Terraform, when running on a particular system, to
+consult only a local filesystem directory where you've created a local mirror
+of the necessary plugins, and to skip accessing the upstream registry at all.
+
+The `terraform providers mirror` command can automatically populate a directory
+that will be used as a local filesystem mirror in the provider installation
+configuration.
+
+-> `terraform providers schema` is available only in Terraform v0.13 or later.
```
Comment (alisdair) [inline website/docs/commands/providers/mirror.html.md:31]:
```suggestion
-> `terraform providers mirror` is available only in Terraform v0.13 or later.
```

**Comment by apparentlymart** [comment]
Heh... thanks for catching those typos @alisdair! I think by the end of the day on Friday I had lost the ability to distinguish those two command lines. I'll fix those before merging.

**Pull request merged**